sql: add index recommendation to EXPLAIN output

nehageorge opened the pull request:
**indexrec: disable inverted index recommendations** 

Previously, we would erroneously give index recommendations of non-
inverted indexes for indexes that should be inverted. In this PR,
we omit index candidates that required inverted indexes. Specifically,
this includes JSON columns, array columns, and spatial data columns.

Release note: None

**sql: add index recommendation to EXPLAIN** 

Previously, the EXPLAIN output only showed the query plan. Since users
often use EXPLAIN to see why a query is poorly performing, it would
be useful for them to also see index recommendations to make the query
faster. In this commit, index recommendations are added to the bottom of
the EXPLAIN output. See the indexrec package under `pkg/sql/opt/indexrec`
to understand how index recommendations are generated.

Release note (sql change): The output of the EXPLAIN SQL statement has changed.
Below the plan, we now output index recommendations for the SQL statement being
"EXPLAIN-ed", if there are any. These index recommendations are indexes the
user could add or indexes they could replace to make the given query faster.

Closes https://github.com/cockroachdb/cockroach/issues/72761.

**indexrec: refactor index recommendation output**

Previously, we manually formatted the string for the index recommendation's
SQL output. This commit takes advantage of the existing functionality for
formatting SQL statements, specifically `CREATE INDEX` and `DROP INDEX`. This
will make it easier to add more types of index recommendations in the future.

Release note: None

Commit:
sql: add index recommendation to EXPLAIN

Previously, the EXPLAIN output only showed the query plan. Since users
often use EXPLAIN to see why a query is poorly performing, it would
be useful for them to also see index recommendations to make the query
faster. In this commit, index recommendations are added to the bottom of
the EXPLAIN output. See the indexrec package under `pkg/sql/opt/indexrec`
to understand how index recommendations are generated.

Release note (sql change): The output of the EXPLAIN SQL statement has changed.
Below the plan, we now output index recommendations for the SQL statement being
"EXPLAIN-ed", if there are any. These index recommendations are indexes the
user could add or indexes they could replace to make the given query faster.

Commit:
indexrec: disable inverted index recommendations

Previously, we would erroneously give index recommendations of non-
inverted indexes for indexes that should be inverted. In this PR,
we omit index candidates that required inverted indexes. Specifically,
this includes candidates with JSON columns, array columns, and
spatial data columns.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/73302)
<!-- Reviewable:end -->

rytaft reviewed (commented):
Nice! I think you're going to need to update a lot of test output...

> Reviewed 8 of 8 files at r1, 10 of 10 files at r2, 6 of 6 files at r3, 2 of 2 files at r4, 3 of 3 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73302#-:-MplTIqL8VWLpiPv_iq0:b32ua1q)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nehageorge)

___
*[pkg/sql/plan_opt.go, line 530 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/73302#-MplQMBAEn6a0_DhUAWR:-MplQMBAEn6a0_DhUAWS:b92g0w1) ([raw file](https://github.com/cockroachdb/cockroach/blob/66227fe1d6f1ecd97b75e6b6b1f931c8586cb1d7/pkg/sql/plan_opt.go#L530)):*
> ```Go
> 	// find potential index candidates in the memo.
> 	if _, isExplain := opc.p.stmt.AST.(*tree.Explain); isExplain {
> 		if err := opc.makeQueryIndexRecommendation(f); err != nil {
> ```

doesn't necessarily need to be in this PR, but make sure that you check a session / cluster setting here so users can optionally disable this feature
___
*[pkg/sql/planner.go, line 228 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/73302#-MplS4-WBsiIOR0HKW4f:-MplS4-WBsiIOR0HKW4g:b33t556) ([raw file](https://github.com/cockroachdb/cockroach/blob/66227fe1d6f1ecd97b75e6b6b1f931c8586cb1d7/pkg/sql/planner.go#L228)):*
> ```Go
> 	// indexRecommendations is a string containing index recommendations for the
> 	// planned statement. This is only set for EXPLAIN statements.
> 	indexRecommendations string
> ```

I think `instrumentationHelper` would be a better place for this. I'd love to avoid adding another field to `planner`
___
*[pkg/sql/opt/indexrec/index_candidate_set.go, line 370 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/73302#-MplPpSCAL3wJF5TROtp:-MplPpSCAL3wJF5TROtq:b-6p57t8) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac86bfc2ceb261a75da22ac372c19601dd893468/pkg/sql/opt/indexrec/index_candidate_set.go#L370)):*
> ```Go
> }
> 
> func requiresInvertedIndex(c *cat.Column) bool {
> ```

It would be good to just reuse this existing function instead if you can: https://github.com/cockroachdb/cockroach/blob/25e38dd31d93509e186d57ef75e798a279e9a5fb/pkg/sql/catalog/colinfo/col_type_info.go#L121
___
*[pkg/sql/opt/testutils/opttester/testdata/index-candidates-recommendations, line 18 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/73302#-MplOK1p-VF0wb2xqTNZ:-MplOK1p-VF0wb2xqTN_:b52d680) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac86bfc2ceb261a75da22ac372c19601dd893468/pkg/sql/opt/testutils/opttester/testdata/index-candidates-recommendations#L18)):*
> ```
> 
> # Ensure that columns that require inverted indexes do not have candidates. This
> # is temporary until we support inverted index recommendations.
> ```

it would be good to also add tests here for predicates that would actually use an inverted index, e.g., 
`j @> '[1]'` or `a @> '{1}'`


<!-- Sent from Reviewable.io -->

nehageorge reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73302)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nehageorge and @rytaft)

___
*[pkg/sql/opt/testutils/opttester/testdata/index-candidates-recommendations, line 18 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/73302#-MplOK1p-VF0wb2xqTNZ:-Mpl_7Q6-LguW0udUIlY:b-tl96zh) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac86bfc2ceb261a75da22ac372c19601dd893468/pkg/sql/opt/testutils/opttester/testdata/index-candidates-recommendations#L18)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

it would be good to also add tests here for predicates that would actually use an inverted index, e.g., 
`j @> '[1]'` or `a @> '{1}'`
</blockquote></details>

I don't know if that would be valuable given my design. Since we don't have rules for `*memo.ContainsExpr` and `*memo.ContainedByExpr`, these types of predicates wouldn't get an index candidate either way. Alternatively, we have a rule for `*memo.IsExpr`, so the fact that there is no candidate here is an indicator that the code is working as intended.


<!-- Sent from Reviewable.io -->

nehageorge reviewed (commented):
Yep, you're right. On it.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73302#-:-MpleUT5CIsioIEAwMkc:b-aws86a)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rytaft)

___
*[pkg/sql/plan_opt.go, line 530 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/73302#-MplQMBAEn6a0_DhUAWR:-MplecON2AwhXLkdghw5:bhinyry) ([raw file](https://github.com/cockroachdb/cockroach/blob/66227fe1d6f1ecd97b75e6b6b1f931c8586cb1d7/pkg/sql/plan_opt.go#L530)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

doesn't necessarily need to be in this PR, but make sure that you check a session / cluster setting here so users can optionally disable this feature
</blockquote></details>

Yep, will do in another PR.
___
*[pkg/sql/opt/indexrec/index_candidate_set.go, line 370 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/73302#-MplPpSCAL3wJF5TROtp:-MplefWN5hPm4BMzezrY:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac86bfc2ceb261a75da22ac372c19601dd893468/pkg/sql/opt/indexrec/index_candidate_set.go#L370)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

It would be good to just reuse this existing function instead if you can: https://github.com/cockroachdb/cockroach/blob/25e38dd31d93509e186d57ef75e798a279e9a5fb/pkg/sql/catalog/colinfo/col_type_info.go#L121
</blockquote></details>

Done.
___
*[pkg/sql/planner.go, line 228 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/73302#-MplS4-WBsiIOR0HKW4f:-Mpleh5a9DNhFVF8YW38:b-j38h0c) ([raw file](https://github.com/cockroachdb/cockroach/blob/66227fe1d6f1ecd97b75e6b6b1f931c8586cb1d7/pkg/sql/planner.go#L228)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

I think `instrumentationHelper` would be a better place for this. I'd love to avoid adding another field to `planner`
</blockquote></details>

This makes sense. Done.


<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
> Reviewed 6 of 6 files at r6, 3 of 3 files at r7, 4 of 4 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73302)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rytaft)

___
*[pkg/sql/opt/testutils/opttester/testdata/index-candidates-recommendations, line 18 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/73302#-MplOK1p-VF0wb2xqTNZ:-MpmlcM34wMMRYHMNQK7:bjmn7b2) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac86bfc2ceb261a75da22ac372c19601dd893468/pkg/sql/opt/testutils/opttester/testdata/index-candidates-recommendations#L18)):*
<details><summary><i>Previously, nehageorge wrote…</i></summary><blockquote>

I don't know if that would be valuable given my design. Since we don't have rules for `*memo.ContainsExpr` and `*memo.ContainedByExpr`, these types of predicates wouldn't get an index candidate either way. Alternatively, we have a rule for `*memo.IsExpr`, so the fact that there is no candidate here is an indicator that the code is working as intended.
</blockquote></details>

True, these cases will mostly be useful once you add support for recommending inverted indexes. Feel free to hold off until then.


<!-- Sent from Reviewable.io -->

Commit:
indexrec: refactor index recommendation output

Previously, we manually formatted the string for the index recommendation's
SQL output. This commit takes advantage of the existing functionality for
formatting SQL statements, specifically `CREATE INDEX` and `DROP INDEX`. This
will make it easier to add more types of index recommendations in the future.

Release note: None

mgartner reviewed (changes requested):
I haven't reviewed the entire PR, but I left a few comments to consider. Looks great so far!

> Reviewed 13 of 13 files at r10, 63 of 63 files at r11.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73302#-:-MqQf2855_Tno_xyQJyo:bo0u3jr)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner, @nehageorge, and @rytaft)

___
*[pkg/sql/plan_opt.go, line 668 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/73302#-MqQOIBs7F_47TI1dYY0:-MqQOIBs7F_47TI1dYY1:b-6hkj0g) ([raw file](https://github.com/cockroachdb/cockroach/blob/60fdd6e9290c6f5f3330b9f164427812404f74b8/pkg/sql/plan_opt.go#L668)):*
> ```Go
> // query is outputted based on which hypothetical indexes are helpful in the
> // optimal plan.
> func (opc *optPlanningCtx) makeQueryIndexRecommendation(f *norm.Factory) error {
> ```

nit: When `f` is passed in as an argument, it's not obvious that this is the same factory in `opc.optimizer`. I think it'd be clearer to have no argument and do `f := opc.optimizer.Factory()` at the top of this function.
___
*[pkg/sql/plan_opt.go, line 673 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/73302#-MqQMg6bDprszZHmPat3:-MqQMg6bDprszZHmPat4:b-u5odcp) ([raw file](https://github.com/cockroachdb/cockroach/blob/60fdd6e9290c6f5f3330b9f164427812404f74b8/pkg/sql/plan_opt.go#L673)):*
> ```Go
> 
> 	// Use the optimizer to fully normalize the memo. The sort enforcer is added
> 	// to the memo in this step.
> ```

Why is `The sort enforcer is added to the memo in this step.` relevant here?
___
*[pkg/sql/plan_opt.go, line 705 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/73302#-MqQOYHTAPMArlYg3rqi:-MqQOYHTAPMArlYg3rqj:b-q1z2bq) ([raw file](https://github.com/cockroachdb/cockroach/blob/60fdd6e9290c6f5f3330b9f164427812404f74b8/pkg/sql/plan_opt.go#L705)):*
> ```Go
> 
> 	// Update the saved memo's metadata with the original table information.
> 	// Prepare to re-optimize and create an executable plan.
> ```

nit: Explain that re-initializing the optimizer also re-initializes the factory.
___
*[pkg/sql/plan_opt.go, line 717 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/73302#-MqQPh9d8n9hAhP2B7mK:-MqQPh9d8n9hAhP2B7mL:bpa5h81) ([raw file](https://github.com/cockroachdb/cockroach/blob/60fdd6e9290c6f5f3330b9f164427812404f74b8/pkg/sql/plan_opt.go#L717)):*
> ```Go
> // hasVirtualTables returns true if there are any virtual tables in the query
> // metadata. Otherwise, it returns false.
> func hasVirtualTables(f *norm.Factory) bool {
> ```

nit: this could also be a method of `optPlanningCtx`.
___
*[pkg/sql/opt/cat/table.go, line 142 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/73302#-MqQQLoL17h2_jaTGVnp:-MqQQLoL17h2_jaTGVnq:b8dac20) ([raw file](https://github.com/cockroachdb/cockroach/blob/60fdd6e9290c6f5f3330b9f164427812404f74b8/pkg/sql/opt/cat/table.go#L142)):*
> ```Go
> 
> 	// LookupColumnOrdinal returns the ordinal of the column with the given ID. A
> 	// cache makes the lookup O(1).
> ```

Adding a function to `cat.Table` that is not used internally by the optimizer, and that involves `descpb.ColumnID` seems like an anti-pattern. I'm going to spend some time to see if some of the code that relies on `cat.Table` in `opt_catalog.go` can be reworked so you don't have to do this.
___
*[pkg/sql/opt/exec/execbuilder/testdata/explain, line 1621 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/73302#-MqQRAvy5PDOkQfNwzv5:-MqQRAvy5PDOkQfNwzv6:b-cfsylf) ([raw file](https://github.com/cockroachdb/cockroach/blob/60fdd6e9290c6f5f3330b9f164427812404f74b8/pkg/sql/opt/exec/execbuilder/testdata/explain#L1621)):*
> ```
> EXPLAIN (OPT, MEMO) SELECT * FROM tc JOIN t ON k=a
> ----
> memo (optimized, ~10KB, required=[presentation: info:10])
> ```

I haven't finished reviewing the PR, but a ~29% reduction in the size of the memo here is certainly surprising. I saw other tests where memos shrank. Do you know what is causing this?
___
*[pkg/sql/opt/indexrec/index_candidate_set.go, line 172 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/73302#-MqQRmitEWRUKs1KHn-t:-MqQRmitEWRUKs1KHn-u:bs1x5em) ([raw file](https://github.com/cockroachdb/cockroach/blob/60fdd6e9290c6f5f3330b9f164427812404f74b8/pkg/sql/opt/indexrec/index_candidate_set.go#L172)):*
> ```Go
> 		if colID == 0 {
> 			continue
> 		}
> ```

I would think it would be impossible for an ordering column to have an ID of `0`. In what case does this occur?
___
*[pkg/sql/opt/indexrec/index_recommendation_set.go, line 153 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/73302#-MqQSEL91DosdgDfRY1X:-MqQSEL91DosdgDfRY1Y:bl3yi5t) ([raw file](https://github.com/cockroachdb/cockroach/blob/60fdd6e9290c6f5f3330b9f164427812404f74b8/pkg/sql/opt/indexrec/index_recommendation_set.go#L153)):*
> ```Go
> 	}
> 
> 	var output []string
> ```

Is it possible to know the number of output lines based on the size of `irs`? If so, you could allocate an exactly-sized slice here with `make`.
___
*[pkg/sql/opt/indexrec/testdata/index-candidates-recommendations, line 25 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/73302#-MqQJv4xApaHpyZ4MbzN:-MqQJv4xApaHpyZ4MbzO:bubx7xl) ([raw file](https://github.com/cockroachdb/cockroach/blob/414c253f2c03fea323f23e36c4823402134b5bd1/pkg/sql/opt/indexrec/testdata/index-candidates-recommendations#L25)):*
> ```
> 
> index-candidates
> SELECT * FROM t4 WHERE k = 2 AND a IS NULL OR f > 2
> ```

nit: use `j` in the query filter, e.g., `j->'a' = '1'`


<!-- Sent from Reviewable.io -->

nehageorge reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73302)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner, @nehageorge, and @rytaft)

___
*[pkg/sql/plan_opt.go, line 673 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/73302#-MqQMg6bDprszZHmPat3:-MqQhaze8Cb9sS3ezYUH:b-w14c4) ([raw file](https://github.com/cockroachdb/cockroach/blob/60fdd6e9290c6f5f3330b9f164427812404f74b8/pkg/sql/plan_opt.go#L673)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Why is `The sort enforcer is added to the memo in this step.` relevant here?
</blockquote></details>

The reason I included it is because otherwise we would not have to call `Optimize` at all prior to determining index candidates. (We would have all the information we need in the `Build` memo).  I can make that more clear in the comment.
___
*[pkg/sql/opt/cat/table.go, line 142 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/73302#-MqQQLoL17h2_jaTGVnp:-MqUgWCi7b-3wtkLDtMx:b-xr6l03) ([raw file](https://github.com/cockroachdb/cockroach/blob/60fdd6e9290c6f5f3330b9f164427812404f74b8/pkg/sql/opt/cat/table.go#L142)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Adding a function to `cat.Table` that is not used internally by the optimizer, and that involves `descpb.ColumnID` seems like an anti-pattern. I'm going to spend some time to see if some of the code that relies on `cat.Table` in `opt_catalog.go` can be reworked so you don't have to do this.
</blockquote></details>

As per our discussion, I will refactor this.
___
*[pkg/sql/opt/exec/execbuilder/testdata/explain, line 1621 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/73302#-MqQRAvy5PDOkQfNwzv5:-MqUgmW-E0R4qa3OVVcf:b-uhkt7m) ([raw file](https://github.com/cockroachdb/cockroach/blob/60fdd6e9290c6f5f3330b9f164427812404f74b8/pkg/sql/opt/exec/execbuilder/testdata/explain#L1621)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

I haven't finished reviewing the PR, but a ~29% reduction in the size of the memo here is certainly surprising. I saw other tests where memos shrank. Do you know what is causing this?
</blockquote></details>

Yea, I believe it comes from the fact that the memo goes through this flow before optimization in addition to the typical flow:
```
savedMemo := opc.optimizer.DetachMemo()
opc.optimizer.Init(f.EvalContext(), &opc.catalog)
if err := f.AssignPlaceholders(savedMemo); err != nil {
		return err
}
```
The reason why this flow causes a difference could potentially warrant further investigation. I thought that it was fine because the code path is different, but maybe this is not expected?
___
*[pkg/sql/opt/indexrec/index_candidate_set.go, line 172 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/73302#-MqQRmitEWRUKs1KHn-t:-MqUc7Be6kkLuhfzHHRW:b-7aa9oh) ([raw file](https://github.com/cockroachdb/cockroach/blob/60fdd6e9290c6f5f3330b9f164427812404f74b8/pkg/sql/opt/indexrec/index_candidate_set.go#L172)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

I would think it would be impossible for an ordering column to have an ID of `0`. In what case does this occur?
</blockquote></details>

IIRC this occurred and was previously causing a test to fail. I am currently trying to figure out which case this happens in. 
___
*[pkg/sql/opt/indexrec/testdata/index-candidates-recommendations, line 25 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/73302#-MqQJv4xApaHpyZ4MbzN:-MqUcRFFCJ0a8zkAbPt6:bigsk9t) ([raw file](https://github.com/cockroachdb/cockroach/blob/414c253f2c03fea323f23e36c4823402134b5bd1/pkg/sql/opt/indexrec/testdata/index-candidates-recommendations#L25)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: use `j` in the query filter, e.g., `j->'a' = '1'`
</blockquote></details>

As I mentioned to Becca previously, this makes sense if we consider the actual use of inverted indexes, but since `FindIndexCandidateSet` doesn't currently match a `*memo.FetchValExpr`, it would not be a candidate in any case. 


<!-- Sent from Reviewable.io -->

nehageorge reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73302)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @rytaft)

___
*[pkg/sql/plan_opt.go, line 668 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/73302#-MqQOIBs7F_47TI1dYY0:-MqVhpibCD0LTICJBuOt:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/60fdd6e9290c6f5f3330b9f164427812404f74b8/pkg/sql/plan_opt.go#L668)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: When `f` is passed in as an argument, it's not obvious that this is the same factory in `opc.optimizer`. I think it'd be clearer to have no argument and do `f := opc.optimizer.Factory()` at the top of this function.
</blockquote></details>

Done.
___
*[pkg/sql/plan_opt.go, line 705 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/73302#-MqQOYHTAPMArlYg3rqi:-MqVho7QBy3a0WHcJ05I:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/60fdd6e9290c6f5f3330b9f164427812404f74b8/pkg/sql/plan_opt.go#L705)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: Explain that re-initializing the optimizer also re-initializes the factory.
</blockquote></details>

Done.
___
*[pkg/sql/plan_opt.go, line 717 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/73302#-MqQPh9d8n9hAhP2B7mK:-MqVhnIcE6f46gHWwNq2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/60fdd6e9290c6f5f3330b9f164427812404f74b8/pkg/sql/plan_opt.go#L717)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: this could also be a method of `optPlanningCtx`.
</blockquote></details>

Done.
___
*[pkg/sql/opt/cat/table.go, line 142 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/73302#-MqQQLoL17h2_jaTGVnp:-MqVhmI533RKQTjEyoc9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/60fdd6e9290c6f5f3330b9f164427812404f74b8/pkg/sql/opt/cat/table.go#L142)):*
<details><summary><i>Previously, nehageorge wrote…</i></summary><blockquote>

As per our discussion, I will refactor this.
</blockquote></details>

Done.
___
*[pkg/sql/opt/indexrec/index_candidate_set.go, line 172 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/73302#-MqQRmitEWRUKs1KHn-t:-MqVh_GlAAn-u_hNIGM0:b-yw30c9) ([raw file](https://github.com/cockroachdb/cockroach/blob/60fdd6e9290c6f5f3330b9f164427812404f74b8/pkg/sql/opt/indexrec/index_candidate_set.go#L172)):*
<details><summary><i>Previously, nehageorge wrote…</i></summary><blockquote>

IIRC this occurred and was previously causing a test to fail. I am currently trying to figure out which case this happens in. 
</blockquote></details>

I couldn't seem to track it down. To speed up the process, I temporarily commented out these lines to see if it breaks anything.
___
*[pkg/sql/opt/indexrec/index_recommendation_set.go, line 153 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/73302#-MqQSEL91DosdgDfRY1X:-MqVhX-uBBblWZJXfNCp:bt1gx32) ([raw file](https://github.com/cockroachdb/cockroach/blob/60fdd6e9290c6f5f3330b9f164427812404f74b8/pkg/sql/opt/indexrec/index_recommendation_set.go#L153)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Is it possible to know the number of output lines based on the size of `irs`? If so, you could allocate an exactly-sized slice here with `make`.
</blockquote></details>

It is, done.


<!-- Sent from Reviewable.io -->

nehageorge reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73302)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @rytaft)

___
*[pkg/sql/opt/indexrec/index_candidate_set.go, line 172 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/73302#-MqQRmitEWRUKs1KHn-t:-MqZc6Al8bVdQnshWtS7:b-w24b6t) ([raw file](https://github.com/cockroachdb/cockroach/blob/60fdd6e9290c6f5f3330b9f164427812404f74b8/pkg/sql/opt/indexrec/index_candidate_set.go#L172)):*
<details><summary><i>Previously, nehageorge wrote…</i></summary><blockquote>

I couldn't seem to track it down. To speed up the process, I temporarily commented out these lines to see if it breaks anything.
</blockquote></details>

Looks like CI is passing. My mistake, I was likely confusing the error of getting a `TableID` of 0 with a `ColumnID` of 0.


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nicely done!

> Reviewed 2 of 13 files at r10, 60 of 60 files at r13, 60 of 60 files at r14, 5 of 5 files at r15, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73302#-:-Mq_tePp81SvXXNPHLUz:b-msncp4)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner, @nehageorge, and @rytaft)

___
*[pkg/sql/plan_opt.go, line 528 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/73302#-Mq_oQHj2Vl4kLoC-YTM:-Mq_oQHj2Vl4kLoC-YTN:brv5tcv) ([raw file](https://github.com/cockroachdb/cockroach/blob/15768481845efe910a0f2cce8702b83ba3fcd544/pkg/sql/plan_opt.go#L528)):*
> ```Go
> 	// For index recommendations, after building we must interrupt the flow to
> 	// find potential index candidates in the memo.
> 	if _, isExplain := opc.p.stmt.AST.(*tree.Explain); isExplain && !opc.hasVirtualTables() {
> ```

Seems like you might want to be able to support index recommendation as long as there is at least one non-virtual table, right? Maybe add a TODO?
___
*[pkg/sql/opt/exec/execbuilder/testdata/union, line 10 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/73302#-Mq_sB1HAGPFtrzKhHbN:-Mq_sB1HAGPFtrzKhHbO:b-jeoh4b) ([raw file](https://github.com/cockroachdb/cockroach/blob/15768481845efe910a0f2cce8702b83ba3fcd544/pkg/sql/opt/exec/execbuilder/testdata/union#L10)):*
> ```
> 
> query T
> EXPLAIN SELECT v FROM uniontest UNION SELECT k FROM uniontest
> ```

In the future we should consider including set operations when determining the index candidates. This query would benefit from one index on k and one index on v, so it could perform a streaming set operation. Obviously not something you should add now, but maybe open an issue?
___
*[pkg/sql/opt/indexrec/index_recommendation_set.go, line 30 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/73302#-Mq_mKkZ7JymRQBPqKFq:-Mq_mKkZ7JymRQBPqKFr:b-lsqqi5) ([raw file](https://github.com/cockroachdb/cockroach/blob/32e29d48c0f3d34741c18f8d19f04c09b0a9d903/pkg/sql/opt/indexrec/index_recommendation_set.go#L30)):*
> ```Go
> // its hypothetical indexes when optimizing, and then prunes the stored columns
> // after, this may result in the best plan not being chosen. Meaning, a plan
> // might not be recommended because the optimizer includes the overhead of
> ```

nit: it's strange to say "a plan is recommended". The optimizer chooses a plan, which determines which indexes are recommended.
___
*[pkg/sql/opt/indexrec/index_recommendation_set.go, line 129 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/73302#-Mq_mp-v3Rutqvyvf5RI:-Mq_mp-v3Rutqvyvf5RJ:b38w7qu) ([raw file](https://github.com/cockroachdb/cockroach/blob/32e29d48c0f3d34741c18f8d19f04c09b0a9d903/pkg/sql/opt/indexrec/index_recommendation_set.go#L129)):*
> ```Go
> 			if index.indexOrdinal == indexOrd {
> 				// Update newStoredColOrds to include all stored column ordinals that
> 				// are in scannedColOrds.
> ```

nit: update comment since you removed `newStoredColOrds`


<!-- Sent from Reviewable.io -->

nehageorge reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73302)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner and @rytaft)

___
*[pkg/sql/plan_opt.go, line 528 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/73302#-Mq_oQHj2Vl4kLoC-YTM:-Mqa6C7N4afIAmKSmeOI:b-19p3al) ([raw file](https://github.com/cockroachdb/cockroach/blob/15768481845efe910a0f2cce8702b83ba3fcd544/pkg/sql/plan_opt.go#L528)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Seems like you might want to be able to support index recommendation as long as there is at least one non-virtual table, right? Maybe add a TODO?
</blockquote></details>

Good point. Moved this check to `addIndexToCandidates` in `index_candidate_set.go` so we avoid candidates from virtual tables.
___
*[pkg/sql/opt/exec/execbuilder/testdata/explain, line 1621 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/73302#-MqQRAvy5PDOkQfNwzv5:-Mqa6R-d6vWl6q4eZ_L6:bx1yz7) ([raw file](https://github.com/cockroachdb/cockroach/blob/60fdd6e9290c6f5f3330b9f164427812404f74b8/pkg/sql/opt/exec/execbuilder/testdata/explain#L1621)):*
<details><summary><i>Previously, nehageorge wrote…</i></summary><blockquote>

Yea, I believe it comes from the fact that the memo goes through this flow before optimization in addition to the typical flow:
```
savedMemo := opc.optimizer.DetachMemo()
opc.optimizer.Init(f.EvalContext(), &opc.catalog)
if err := f.AssignPlaceholders(savedMemo); err != nil {
		return err
}
```
The reason why this flow causes a difference could potentially warrant further investigation. I thought that it was fine because the code path is different, but maybe this is not expected?


</blockquote></details>

Added a function to copy the memo's size when copying the memo, to avoid this discrepancy. 
___
*[pkg/sql/opt/exec/execbuilder/testdata/union, line 10 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/73302#-Mq_sB1HAGPFtrzKhHbN:-Mqa6fe2EF5gTH1GP9_T:bm00d2l) ([raw file](https://github.com/cockroachdb/cockroach/blob/15768481845efe910a0f2cce8702b83ba3fcd544/pkg/sql/opt/exec/execbuilder/testdata/union#L10)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

In the future we should consider including set operations when determining the index candidates. This query would benefit from one index on k and one index on v, so it could perform a streaming set operation. Obviously not something you should add now, but maybe open an issue?
</blockquote></details>

Will create an issue.
___
*[pkg/sql/opt/indexrec/index_recommendation_set.go, line 30 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/73302#-Mq_mKkZ7JymRQBPqKFq:-Mqa6jF7CK6VX9qW5kRG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/32e29d48c0f3d34741c18f8d19f04c09b0a9d903/pkg/sql/opt/indexrec/index_recommendation_set.go#L30)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

nit: it's strange to say "a plan is recommended". The optimizer chooses a plan, which determines which indexes are recommended.
</blockquote></details>

Done.
___
*[pkg/sql/opt/indexrec/index_recommendation_set.go, line 129 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/73302#-Mq_mp-v3Rutqvyvf5RI:-Mqa6kzT86BhRNEhnsRH:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/32e29d48c0f3d34741c18f8d19f04c09b0a9d903/pkg/sql/opt/indexrec/index_recommendation_set.go#L129)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

nit: update comment since you removed `newStoredColOrds`
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
> Reviewed 8 of 8 files at r16, 1 of 3 files at r18, 2 of 2 files at r19, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73302)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner, @nehageorge, and @rytaft)

___
*[pkg/sql/plan_opt.go, line 702 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/73302#-MqaBmiI899GDnpW1igq:-MqaBmiK-OUDULJllzSm:b-rb38ur) ([raw file](https://github.com/cockroachdb/cockroach/blob/b10617e432c3287cc415d3542ad5adb444e33ef3/pkg/sql/plan_opt.go#L702)):*
> ```Go
> 		savedMemo.RootProps(),
> 		replaceNoopFn,
> 	)
> ```

Why did you change this?

If this is needed, seems worth turning this into another function called `Copy` or `CopyMemo`, with `replaceNoopFn` defined in `factory.go` (or instead of `replaceNoopFn` just use -- and possibly rename -- the existing function `CopyScalarWithoutPlaceholders` that does the same thing)
___
*[pkg/sql/opt/indexrec/index_candidate_set.go, line 348 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/73302#-MqaDTj-2sXBPQLqZOCX:-MqaDTj-2sXBPQLqZOCY:b-wf0zo8) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0f76049166ee2df003e1ef7dc21fa9b87b949cf/pkg/sql/opt/indexrec/index_candidate_set.go#L348)):*
> ```Go
> ) {
> 	// Do not add candidates from virtual tables.
> 	if currTable.IsVirtualTable() {
> ```

should you add a test for a query that uses one virtual table and one non-virtual?
___
*[pkg/sql/opt/norm/factory.go, line 255 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/73302#-MqaDG1h3W8PSrFqmlj7:-MqaDG1h3W8PSrFqmlj8:b8y4idq) ([raw file](https://github.com/cockroachdb/cockroach/blob/b10617e432c3287cc415d3542ad5adb444e33ef3/pkg/sql/opt/norm/factory.go#L255)):*
> ```Go
> 
> 	// Rewrite the memory estimate to get rid of potential discrepancies.
> 	f.mem.CopyMemEstimateFrom(from.Memo())
> ```

Why do we need this? If the new size is smaller then shouldn't we be using the new size? I think certain things are removed from the memo when it's detached and copied.


<!-- Sent from Reviewable.io -->

nehageorge reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73302)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner and @rytaft)

___
*[pkg/sql/plan_opt.go, line 702 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/73302#-MqaBmiI899GDnpW1igq:-MqaF1Aw42GP0VUNmYm8:bn54r7t) ([raw file](https://github.com/cockroachdb/cockroach/blob/b10617e432c3287cc415d3542ad5adb444e33ef3/pkg/sql/plan_opt.go#L702)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Why did you change this?

If this is needed, seems worth turning this into another function called `Copy` or `CopyMemo`, with `replaceNoopFn` defined in `factory.go` (or instead of `replaceNoopFn` just use -- and possibly rename -- the existing function `CopyScalarWithoutPlaceholders` that does the same thing)
</blockquote></details>

I was discussing with Marcus via Slack and he brought up that `AssignPlaceholders` could be misleading as that isn't what is being done. So this was an alternative. I renamed the function to `CopyScalarNoop`.
___
*[pkg/sql/opt/indexrec/index_candidate_set.go, line 348 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/73302#-MqaDTj-2sXBPQLqZOCX:-MqaPnHy4YLAXZEzTMf_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0f76049166ee2df003e1ef7dc21fa9b87b949cf/pkg/sql/opt/indexrec/index_candidate_set.go#L348)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

should you add a test for a query that uses one virtual table and one non-virtual?
</blockquote></details>

Done.
___
*[pkg/sql/opt/norm/factory.go, line 255 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/73302#-MqaDG1h3W8PSrFqmlj7:-MqaPobR2evmaLyOAzdS:b-j8uogx) ([raw file](https://github.com/cockroachdb/cockroach/blob/b10617e432c3287cc415d3542ad5adb444e33ef3/pkg/sql/opt/norm/factory.go#L255)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Why do we need this? If the new size is smaller then shouldn't we be using the new size? I think certain things are removed from the memo when it's detached and copied.
</blockquote></details>

Oh that makes sense. I thought it was an incorrect computation. Removed this function.


<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
> Reviewed 3 of 7 files at r20, 1 of 3 files at r21, 3 of 3 files at r22, 2 of 2 files at r23, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73302)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner, @nehageorge, and @rytaft)

___
*[pkg/sql/plan_opt.go, line 702 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/73302#-MqaBmiI899GDnpW1igq:-Mqa_Ab-9QImqnumF8CC:bsxd3yb) ([raw file](https://github.com/cockroachdb/cockroach/blob/b10617e432c3287cc415d3542ad5adb444e33ef3/pkg/sql/plan_opt.go#L702)):*
<details><summary><i>Previously, nehageorge wrote…</i></summary><blockquote>

I was discussing with Marcus via Slack and he brought up that `AssignPlaceholders` could be misleading as that isn't what is being done. So this was an alternative. I renamed the function to `CopyScalarNoop`.
</blockquote></details>

How about `CopyExprNoop`? This memo will include non-scalar expressions, and there's no reason to limit the function to scalars since it operates on `opt.Expr`s 


<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73302)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner, @nehageorge, and @rytaft)

___
*[pkg/sql/plan_opt.go, line 702 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/73302#-MqaBmiI899GDnpW1igq:-MqaammL0LecyufbrvqJ:b-mjhhn) ([raw file](https://github.com/cockroachdb/cockroach/blob/b10617e432c3287cc415d3542ad5adb444e33ef3/pkg/sql/plan_opt.go#L702)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

How about `CopyExprNoop`? This memo will include non-scalar expressions, and there's no reason to limit the function to scalars since it operates on `opt.Expr`s 
</blockquote></details>

Or to be more explicit `CopyWithoutAssigningPlaceholders`. I think the "Scalar" piece was mostly what was distracting me. Anyway, up to you.


<!-- Sent from Reviewable.io -->

nehageorge reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73302)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner and @rytaft)

___
*[pkg/sql/plan_opt.go, line 702 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/73302#-MqaBmiI899GDnpW1igq:-MqoPolF89_QBkUHVLEJ:b-cgwwdm) ([raw file](https://github.com/cockroachdb/cockroach/blob/b10617e432c3287cc415d3542ad5adb444e33ef3/pkg/sql/plan_opt.go#L702)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Or to be more explicit `CopyWithoutAssigningPlaceholders`. I think the "Scalar" piece was mostly what was distracting me. Anyway, up to you.
</blockquote></details>

Makes sense. I renamed the function.


<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 7 of 7 files at r24, 2 of 2 files at r25, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73302#-:-MqoYycX2ymDcWAUSL0x:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner and @rytaft)



<!-- Sent from Reviewable.io -->

mgartner reviewed (approved):
Great work!

> Reviewed 60 of 60 files at r13, 51 of 60 files at r14, 1 of 8 files at r16, 1 of 7 files at r20, 7 of 7 files at r24, 2 of 2 files at r25, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73302#-:-MqozXWA1LyKhP8IJ0M4:b-9dwnzc)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner and @rytaft)

___
*[pkg/sql/opt/norm/factory.go, line 255 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/73302#-MqaDG1h3W8PSrFqmlj7:-Mqowhb9EBXVbS9yfZHt:bf7gruq) ([raw file](https://github.com/cockroachdb/cockroach/blob/b10617e432c3287cc415d3542ad5adb444e33ef3/pkg/sql/opt/norm/factory.go#L255)):*
<details><summary><i>Previously, nehageorge wrote…</i></summary><blockquote>

Oh that makes sense. I thought it was an incorrect computation. Removed this function.
</blockquote></details>

I suggested this because the memo size in `EXPLAIN (MEMO, VERBOSE) ...` statements changed. Previously there was no detachment and copy step in `EXPLAIN`, so the `memEstimate` included sizes of expressions added to metadata. Now that we are copying the memo, `memEstimate` is reset to `0`, and the `EXPLAIN (MEMO, VERBOSE) ...` memo size does not include the size of the metadata expressions. I believe this was causing the memo sizes in tests to change.

But I'm not sure how important this is, and my intuition is that it's not critical to have the previous behavior, so maybe we leave as-is?


<!-- Sent from Reviewable.io -->

nehageorge reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73302)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner and @rytaft)

___
*[pkg/sql/opt/norm/factory.go, line 255 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/73302#-MqaDG1h3W8PSrFqmlj7:-Mqp3iIL0h91zCU6PR5Q:b7f4imo) ([raw file](https://github.com/cockroachdb/cockroach/blob/b10617e432c3287cc415d3542ad5adb444e33ef3/pkg/sql/opt/norm/factory.go#L255)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

I suggested this because the memo size in `EXPLAIN (MEMO, VERBOSE) ...` statements changed. Previously there was no detachment and copy step in `EXPLAIN`, so the `memEstimate` included sizes of expressions added to metadata. Now that we are copying the memo, `memEstimate` is reset to `0`, and the `EXPLAIN (MEMO, VERBOSE) ...` memo size does not include the size of the metadata expressions. I believe this was causing the memo sizes in tests to change.

But I'm not sure how important this is, and my intuition is that it's not critical to have the previous behavior, so maybe we leave as-is?
</blockquote></details>

Makes sense. After evaluating the situation further, it seems like it's not critical to preserve the previous behaviour. Also, manually overwriting the memo size seems like it might be unnecessary. I'm going to leave it as is.


<!-- Sent from Reviewable.io -->

nehageorge commented:
TFTRs!!

bors r+
